send/buffered.go

```
@@ -2,134 +2,158 @@ package send
+// when the buffer reaches a specified size or or after a specified interval
+// has passed.
 //
-// If the interval is 0, the constructor sets an interval of 1 minute, and if
-// it is less than 5 seconds, the constructor sets it to 5 seconds. If the
-// size threshold is 0, then the constructor sets a threshold of 100.
+// If the flushInterval is 0, the constructor sets an flushInterval of 1 minute.
+// If the flushInterval is less than 5 seconds, the constructor sets it to 5 seconds.
+// If the bufferSize threshold is 0, the constructor sets a threshold of 100.
 //
 // This Sender does not own the underlying Sender, so users are responsible for
 // closing the underlying Sender if/when it is appropriate to release its
 // resources.
-func NewBufferedSender(sender Sender, interval time.Duration, size int) Sender {
-	if interval == 0 {
-		interval = time.Minute
-	} else if interval < minInterval {
-		interval = minInterval
-	}
-
-	if size <= 0 {
-		size = 100
-	}
+func NewBufferedSender(ctx context.Context, sender Sender, opts BufferedSenderOptions) Sender {
+	opts.validate()
 
-	ctx, cancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithCancel(ctx)
 	s := &bufferedSender{
-		Sender: sender,
-		cancel: cancel,
-		buffer: []message.Composer{},
-		size:   size,
+		Sender:     sender,
+		ctx:        ctx,
+		cancel:     cancel,
+		opts:       opts,
+		buffer:     make([]message.Composer, 0, opts.BufferSize),
+		needsFlush: make(chan bool, 1),
+		incoming:   make(chan message.Composer, incomingBufferFactor*opts.BufferSize),
 	}
 
-	go s.intervalFlush(ctx, interval)
+	go s.consumer()
 
 	return s
 }
 
+type BufferedSenderOptions struct {
+	FlushInterval time.Duration
+	BufferSize    int
+}
+
+func (opts *BufferedSenderOptions) validate() {
+	if opts.FlushInterval == 0 {
+		opts.FlushInterval = time.Minute
+	} else if opts.FlushInterval < minInterval {
+		opts.FlushInterval = minInterval
+	}
+
+	if opts.BufferSize <= 0 {
```

I would return an error if `BufferSize` is negative, since that's not a valid number.

---

done